### PR TITLE
Implemented vines

### LINKED ISSUE #572: Implement vines

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Closes #572.

### CODE REVIEW

File: server/block/vine.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+// FlammabilityInfo ...
+func (Vines) FlammabilityInfo() FlammabilityInfo {
+	return newFlammabilityInfo(15, 100, true)
+}
+
+// BreakInfo ...
+func (v Vines) BreakInfo() BreakInfo {
+	return newBreakInfo(0.2, alwaysHarvestable, shearsEffective, oneOf(v))
```
Comment (DaPigGuy) [inline server/block/vine.go:35]: Not always harvestable, requires shears or a silk touch axe

Comment (JustTalDevelops) [reply server/block/vine.go:35]: Any ideas on how I should check for silk touch axes?

Comment (JustTalDevelops) [reply server/block/vine.go]: Updated.

File: server/block/vine.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		}
+
+		_, air := w.Block(selectedPos).(Air)
+		newVines, vines := w.Block(selectedPos).(Vines)
+		if face == cube.FaceUp {
+			if !air {
+				// Vines can't spread upwards unless there is air.
+				return
+			}
+
```
Comment (DaPigGuy) [inline server/block/vine.go:156]: According to the wiki, there can not be more than 4 vines in a 9x9x3 area to spread upwards

File: server/block/vine.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+			for _, f := range cube.HorizontalFaces() {
+				if r.Intn(2) == 0 && v.Attachment(f.Direction()) {
+					newVines, changed = newVines.SetAttachment(f.Direction(), true), true
+				}
+			}
+			if changed {
+				w.SetBlock(selectedPos, newVines, nil)
+			}
+			return
+		}
```
Comment (DaPigGuy) [inline server/block/vine.go:179]: Please check this behavior as it does not match what the wiki says for downwards spread to non-air blocks

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/vine.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+	place(w, pos, v, user, ctx)
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick ...
+func (v Vines) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	above, updated := w.Block(pos.Side(cube.FaceUp)), false
+	for _, d := range v.Attachments() {
+		if _, ok := w.Block(pos.Side(d.Face())).Model().(model.Solid); !ok {
+			if o, ok := above.(Vines); !ok || ok && !o.Attachment(d) {
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/block/vine.go:117]: What about it?

File: server/block/vine.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+// EntityInside ...
+func (Vines) EntityInside(_ cube.Pos, _ *world.Tx, e world.Entity) {
+	if fallEntity, ok := e.(fallDistanceEntity); ok {
+		fallEntity.ResetFallDistance()
+	}
+}
+
+// SetAttachment sets an attachment on the given cube.Direction.
+func (v Vines) SetAttachment(direction cube.Direction, attached bool) Vines {
```
Comment (Sandertv) [inline server/block/vine.go:69]: Would name this `WithAttachment` to make it clear that this does not modify the original block.

**Review by Sandertv — COMMENTED** [review]
Looks good! Got one comment on a method that should probably be renamed, but otherwise no major concerns. Feel free to merge if nothing is left to do and all is tested.

**Pull request merged**